### PR TITLE
add hack dir

### PR DESCRIPTION
**Pull request by ldpliu**

### CODE REVIEW

File: hack/update-apiserver-gen.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+# licensed Materials - Property of IBM
```
Comment (qiujian16) [inline hack/update-apiserver-gen.sh:3]: license not correct

File: Makefile [hunk 2]
```diff
@@ -24,6 +24,11 @@ IMAGE_REGISTRY ?= quay.io/open-cluster-management
 IMAGE_TAG ?= latest
 IMAGE_NAME ?= $(IMAGE_REGISTRY)/$(IMAGE):$(IMAGE_TAG)
 
+GIT_HOST ?= github.com/open-cluster-management
+BASE_DIR := $(shell basename $(PWD))
+DEST := $(GOPATH)/src/$(GIT_HOST)/$(BASE_DIR)
+BINDIR ?= output
```
Comment (qiujian16) [inline Makefile:30]: maybe _output

File: Makefile [hunk 3]
```diff
@@ -72,6 +77,55 @@ build-e2e:
...
+	go build -o $@ $(DEST)/vendor/k8s.io/code-generator/cmd/openapi-gen
+
+$(BINDIR)/go-to-protobuf:
+	go build -o $@ $(DEST)/vendor/k8s.io/code-generator/cmd/go-to-protobuf
+
+$(BINDIR)/protoc-gen-gogo:
+	go build -o $@ $(DEST)/vendor/k8s.io/code-generator/cmd/go-to-protobuf/protoc-gen-gogo
+
+# Regenerate all files if the gen exes changed or any "types.go" files changed
+generate_files: generate_exes $(TYPES_FILES)
```
Comment (qiujian16) [inline Makefile:122]: do we need this TYPES_FILES?

File: Makefile [hunk 4]
```diff
@@ -72,6 +77,55 @@ build-e2e:
...
+$(BINDIR)/defaulter-gen:
+	go build -o $@ $(DEST)/vendor/k8s.io/code-generator/cmd/defaulter-gen
+
+$(BINDIR)/deepcopy-gen:
+	go build -o $@ $(DEST)/vendor/k8s.io/code-generator/cmd/deepcopy-gen
+
+$(BINDIR)/conversion-gen:
+	go build -o $@ $(DEST)/vendor/k8s.io/code-generator/cmd/conversion-gen
+
+$(BINDIR)/client-gen:
```
Comment (qiujian16) [inline Makefile:103]: i do not think we need client/lister/informer

File: hack/update-protobuf.sh [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+install the platform appropriate Protobuf package for your OS:
+  https://github.com/google/protobuf/releases"
+  exit 1
+fi
+
+PATH="$PATH:$BINDIR" go-to-protobuf \
+  --output-base="${GOPATH}/src" \
+  --apimachinery-packages='-k8s.io/apimachinery/pkg/util/intstr,-k8s.io/apimachinery/pkg/api/resource,-k8s.io/apimachinery/pkg/runtime/schema,-k8s.io/apimachinery/pkg/runtime,-k8s.io/apimachinery/pkg/apis/meta/v1,-k8s.io/apimachinery/pkg/apis/meta/v1beta1,-k8s.io/api/core/v1,-k8s.io/api/rbac/v1,-k8s.io/api/certificates/v1beta1' \
+  --go-header-file="${REPO_ROOT}"/hack/custom-boilerplate.go.txt \
+  --proto-import="${REPO_ROOT}"/third_party/protobuf \
```
Comment (qiujian16) [inline hack/update-protobuf.sh:48]: I  do not think we have third_party

File: Makefile [hunk 6]
```diff
@@ -72,6 +77,55 @@ build-e2e:
...
+$(BINDIR)/informer-gen:
+	go build -o $@ $(DEST)/vendor/k8s.io/code-generator/cmd/informer-gen
+
+$(BINDIR)/openapi-gen:
+	go build -o $@ $(DEST)/vendor/k8s.io/code-generator/cmd/openapi-gen
+
+$(BINDIR)/go-to-protobuf:
+	go build -o $@ $(DEST)/vendor/k8s.io/code-generator/cmd/go-to-protobuf
+
+$(BINDIR)/protoc-gen-gogo:
```
Comment (qiujian16) [inline Makefile:118]: is this really needed?

Comment (ldpliu) [reply Makefile:106]: it's needed for generate protebuf

File: hack/update-protobuf.sh (hunk 5 above)
Comment (ldpliu) [reply hack/update-protobuf.sh]: it's used to generate protebuf

**Pull request merged**